segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java

```diff
@@ -331,26 +322,23 @@ public String toString() {
     @Override
     public void add(SegmentOperation operation) throws DataCorruptionException {
         ensureInitializedAndNotClosed();
+        if (!(operation instanceof StorageOperation)) {
```

Comment:
Between this and the Javadoc at the top of the file it seems like maybe a refactoring is due.

Reply:
Well this was a refactoring. The `SegmentAggregator` deals exclusively with transferring data within the segment and the `AttributeAggregator` handles attribute transfer.

I decided to keep segment deletion logic in here because we want to remember to delete attributes when we delete the segment, when we merge it, and one more complex situation. Instead of doing it twice, I preferred to keep that part in one place.

I don't understand what is wrong with this line though. The `SegmentAggregator` only handles `StorageOperation`s; all other types of operations are `MetadataOperation`s are not its responsibility. There are other processors for each segment within the StorageWriter, and each deals with its own types of operations. `WriterTableProcessor` only handles append operations, `AttributeAggregator` handles `AttributeUpdaterOperations` (appends & attribute updates) and this handles appends, seals, mergers, deletes, truncates, etc. In other words, anything which applies directly to Tier 2 is a `StorageOperation` and is handled by this class, while others are a bit more specialized and do not touch Tier 2 (directly).

Reply:
Let's see if I can explain. The `StorageWriter` picks Operations (all of them) from Tier1/DurableLog. The `StorageWriter` also has `WriterProcessors` for each active segment in the container. A `WriterProcessor` has a `SegmentAggregator`, an `AttributeAggregator` (new in this PR) and a `WriterTableProcessor` (for Table Segments).

The `StorageWriter` picks operation from Tier 1 (in order) and sends them to all its `WriterProcessor`s, which in turn sends them to all of the aggregators sub-processors. So from this POV, it is using the visitor pattern (in a way). The operations themselves are immutable, but each Processor/Aggregator knows which operations it needs to process and will just ignore the rest. IMO, this can be done in two ways: the current way, in which all processors see all operations and decide which ones they use, and another in which they register with the StorageWriter to send them only those operations they care about. I just happen to do the former, because even if it accepts an operation of a given type, it may end up ignoring it due to other factors.

Reply:
That sounds reasonable. 
Alternatively: You could also make WriterProcessor the one to send some types to the SegmentAddregator and others to the WriterTableProcessor. (Obviously this makes it more aware of what those classes are doing, so its up to you as to whether that falls too far outside of its current responsibilities.)

Reply:
Good suggestion. Since this is outside of the scope of this change, I captured this in https://github.com/pravega/pravega/issues/4493, which also includes an additional refactoring of the StorageWriter to make it more efficient.